content/posts/replacing-the-redisgraph-parser.md

```diff
@@ -0,0 +1,109 @@
+
+To use [Martin Fowler's language of technical debt](https://martinfowler.com/bliki/TechnicalDebt.html), our self-maintained grammar charged us interest every time we introduced or modified a syntactical construct. Gradual improvement was not an option, however, so we were reluctant to undertake the effort of a full replacement. We finally relented when working on the WITH clause, which is how Cypher allows for multiple queries to be chained into a single atomic execution. Unlike subqueries in SQL, which create a tree of independent executions, the WITH clause flattens query parts into a linear composition:
+
+![Multi-Part Query Railroad Diagram](/multipart_query.png)
+[Cypher railroad diagram for multi-part queries](https://s3.amazonaws.com/artifacts.opencypher.org/railroad/MultiPartQuery.html)
+
+Our original grammar was effective at interpreting a sequence of clauses as a self-contained query, but it would have required a major rewrite to interpret that same sequence as one scope capable of reading from and projecting into others. libcypher-parser had already solved this problem and many more, so we started the even more ambitious rewrite of scrapping our parser and building off an open-source project.
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
+
+RedisGraph is a better project for the issues opened, features requested, and contributions made by our open-source users. While integrating libcypher-parser, we've contributed code and feedback upstream, and collaboration that improves both projects is ongoing.
+
+The lesson that a vibrant open-source community enriches us all is not a new one, but I always find it heartening to be reminded of it!
+
+As an aside, it is interesting to note that instead of building a LALR parser (as Lemon does), libcypher-parser uses a PEG grammar. I won't go into the implications of this here, but Guido van Rossum has been writing a [very interesting series](https://medium.com/@gvanrossum_83706/peg-parsers-7ed72462f97c) on the subject for the curious.
+
+## Paid back with interest
+From start to finish, this undertaking took about 3 months, though even now we have open PRs to better leverage the new system. We now work with an immutable (though annotatable) AST, and have radically modified our access patterns and AST translation layer.
+
+To torture the metaphor of technical debt a little more, we recouped more than the principal over the course of this refactor. libcypher-parser gave us automatic support for a  number of features, ranging from syntactic sugar like implicit edges and partially-specified traversal ranges:
+```console
+MATCH (node1)-->(node2), (src)-[*2..]->(two_or_more_hops)
+```
+To more robust query validations, built-in precedence management for operators and comparisons, and handling for more edge cases than I can recall.
```

Comment:
Can you rephrase this? Is this trying to convey that built-in precedence management for operators and comparisons handle more robust query validations and edge cases?